pkg/wakaama/contrib/objects/on_off_switch.c

```diff
@@ -0,0 +1,492 @@
+            (*data_array)[i].id = res_list[i];
+        }
+    }
+
+    /* now get the values */
+    i = 0;
+    do {
+        DEBUG("[lwm2m:on_off_switch:read]: reading resource %" PRId16 "\n", (*data_array)[i].id);
+        result = _get_value(&(*data_array)[i], instance);
+        i++;
+    } while (i < *num_data && COAP_205_CONTENT == result);
+
+free_out:
+    mutex_unlock(&_on_off_switch_object.lock);
+    return result;
+}
+
+static uint8_t _write_cb(lwm2m_context_t * context, uint16_t instance_id, int num_data,
+                         lwm2m_data_t * data_array, lwm2m_object_t * object, lwm2m_write_type_t write_type)
+{
+    (void)context;
+    (void)write_type;
+
+    lwm2m_obj_on_off_switch_inst_t *instance;
+    uint8_t result = COAP_204_CHANGED;
+
+    mutex_lock(&_on_off_switch_object.lock);
+
+    /* try to get the requested instance from the object list */
+    instance = (lwm2m_obj_on_off_switch_inst_t *)lwm2m_list_find(object->instanceList, instance_id);
+    if (!instance) {
+        DEBUG("[lwm2m:on_off_switch:write]: can't find instance %" PRId16 "\n", instance_id);
+        result = COAP_404_NOT_FOUND;
+        goto free_out;
+    }
+
+    for (int i = 0; i < num_data && result == COAP_204_CHANGED; i++) {
+        switch (data_array[i].id) {
+        case LWM2M_ON_OFF_SWITCH_DIGITAL_INPUT_STATE_ID: {
+            bool prev_status = instance->status;
+
+            lwm2m_data_decode_bool(&data_array[i], &instance->status);
+
+            /* reset timer on transitions */
+            if (instance->status != prev_status) {
+                ztimer_stopwatch_reset(&instance->stopwatch);
+
+                if (instance->status) {
+                    instance->counter++;
+                }
+            }
+
+            _mark_resource_changed(instance_id, LWM2M_ON_OFF_SWITCH_DIGITAL_INPUT_STATE_ID);
+            break;
+        }
+
+        case LWM2M_ON_OFF_SWITCH_ON_TIME_ID:
+        {
+            int64_t val;
+            lwm2m_data_decode_int(&data_array[i], &val);
+            if (val != 0) {
+                DEBUG("[lwm2m:on_off_switch:write]: invalid on_time value, only can write 0\n");
+                result = COAP_400_BAD_REQUEST;
+            } else {
+                ztimer_stopwatch_reset(&instance->stopwatch);
+                _mark_resource_changed(instance_id, LWM2M_ON_OFF_SWITCH_ON_TIME_ID);
+            }
+            break;
+        }
+
+        case LWM2M_ON_OFF_SWITCH_OFF_TIME_ID:
+        {
+            int64_t val;
+            lwm2m_data_decode_int(&data_array[i], &val);
+            if (val != 0) {
+                DEBUG("[lwm2m:on_off_switch:write]: invalid off_time value, only can write 0\n");
+                result = COAP_400_BAD_REQUEST;
+            } else {
+                ztimer_stopwatch_reset(&instance->stopwatch);
+                _mark_resource_changed(instance_id, LWM2M_ON_OFF_SWITCH_OFF_TIME_ID);
+            }
+            break;
+        }
+
+        case LWM2M_ON_OFF_SWITCH_APP_TYPE_ID:
```

Comment:
Could you clarify to me why we have this functionality here and lower in the `lwm2m_object_on_off_switch_update_app_type()`? It's not a drop-in replacement due to the mutex tho.

Reply:
I changed the user functions to use the new `_set_value` function.